internal/pkg/agent/application/actions/handlers/handler_helpers.go

```diff
@@ -0,0 +1,174 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package handlers
+
+import (
+	"context"
+	"time"
+
+	"golang.org/x/sync/errgroup"
+
+	"github.com/elastic/elastic-agent-client/v7/pkg/client"
+	"github.com/elastic/elastic-agent-libs/logp"
+
+	"github.com/elastic/elastic-agent/internal/pkg/agent/application/coordinator"
+	"github.com/elastic/elastic-agent/internal/pkg/agent/errors"
+	"github.com/elastic/elastic-agent/internal/pkg/core/backoff"
+	"github.com/elastic/elastic-agent/internal/pkg/fleetapi"
+	"github.com/elastic/elastic-agent/pkg/component"
+	"github.com/elastic/elastic-agent/pkg/component/runtime"
+)
+
+type actionCoordinator interface {
+	State() coordinator.State
+	PerformAction(ctx context.Context, comp component.Component, unit component.Unit, name string, params map[string]interface{}) (map[string]interface{}, error)
+}
+
+type upgradeCoordinator interface {
+	actionCoordinator
+	Upgrade(ctx context.Context, version string, sourceURI string, action *fleetapi.ActionUpgrade, skipVerifyOverride bool, pgpBytes ...string) error
+}
+
+type performActionFunc func(context.Context, component.Component, component.Unit, string, map[string]interface{}) (map[string]interface{}, error)
+
+type dispatchableAction interface {
+	MarshalMap() (map[string]interface{}, error)
+	Type() string
+}
+
+type unitWithComponent struct {
+	unit      component.Unit
+	component component.Component
+}
+
+func findMatchingUnitsByActionType(state coordinator.State, typ string) []unitWithComponent {
+	ucs := make([]unitWithComponent, 0)
+	for _, comp := range state.Components {
+		if comp.Component.InputSpec != nil && contains(comp.Component.InputSpec.Spec.ProxiedActions, typ) {
+			name := comp.Component.InputType
+			for _, unit := range comp.Component.Units {
+				// All input units should match the component input type, but let's be cautious
+				if unit.Type == client.UnitTypeInput && unit.Config != nil && unit.Config.Type == name {
+					ucs = append(ucs, unitWithComponent{unit, comp.Component})
+				}
+			}
+		}
+	}
+	return ucs
+}
+
+func contains[T comparable](arr []T, val T) bool {
+	for _, v := range arr {
+		if v == val {
+			return true
+		}
+	}
+	return false
+}
+
+type proxiedActionsNotifier struct {
+	log           *logp.Logger
+	performAction performActionFunc
+
+	timeout    time.Duration
+	minBackoff time.Duration
+	maxBackoff time.Duration
+}
+
+const (
+	defaultActionDispatcherTimeout    = 40 * time.Second
+	defaultActionDispatcherBackoffMin = 500 * time.Millisecond
+	defaultActionDispatcherBackoffMax = 10 * time.Second
```

Comment:
These could all be configurable in the agent policy, this would allow us to change them using the new Fleet agent policy override API if we find out they are too short for whatever reason.

For now, disabling the tampering feature flag is also an escape hatch for this problem.